drivers/wdf/wdfldr/common.c

```diff
@@ -0,0 +1,826 @@
+	{
+		goto clean;
+	}
+
+	if (*pInfoLength < modulesSize) 
+	{
+		status = STATUS_BUFFER_TOO_SMALL;
+		goto end;
+	}
+
+	if (pSysInfo->NumberOfModules == 0)
+		goto end;
+
+	for (index = 0; index < pSysInfo->NumberOfModules; ++index, ModuleInfo++) 
+	{
+		ModuleInfo->BasicInfo.ImageBase = pSysInfo->Modules[index].ImageBase;
+		ModuleInfo->ImageSize = pSysInfo->Modules[index].ImageSize;
+		ModuleInfo->FileNameOffset = pSysInfo->Modules[index].OffsetToFileName;
+		memcpy(ModuleInfo->FullPathName, pSysInfo->Modules[index].FullPathName, 0x100);
+	}
+
+end:
+	*pInfoLength = modulesSize;
+clean:
+	if (pSysInfo != &systemInformation)
+		ExFreePoolWithTag(pSysInfo, 0);
+
+	return status;
+}
+
+
+NTSTATUS
+NTAPI
+GetImageBase(
+	IN PCUNICODE_STRING ImageName,
+	OUT PVOID* ImageBase,
+	OUT PULONG ImageSize
+)
+{
+	PCHAR fileName;
+	PRTL_MODULE_EXTENDED_INFO pModuleInfo;
+	PVOID infoBuffer;
+	ULONG_PTR endOfArray;
+	STRING ansiImageName;
+	size_t fileNameLength;
+	ULONG informationLength;
+	ULONG totalSize;
+	ULONG numberOfBytes;
+	NTSTATUS status;
+
+	infoBuffer = NULL;
+	*ImageBase = 0;
+	*ImageSize = 0;
+	ansiImageName.Length = 0;
+	ansiImageName.MaximumLength = 0;
+	ansiImageName.Buffer = 0;
+	informationLength = 0;
+	pModuleInfo = NULL;
+	status = RtlUnicodeStringToAnsiString(&ansiImageName, ImageName, TRUE);
+
+	if (NT_SUCCESS(status) && ansiImageName.Buffer) 
+	{
+		ansiImageName.Buffer[ansiImageName.Length] = 0;
+		fileName = GetFileName(ansiImageName.Buffer);
+		fileNameLength = strlen(fileName);
+
+		if (fileName == NULL ||	fileNameLength == 0)
+		{
+			status = STATUS_OBJECT_NAME_NOT_FOUND;
+			goto clean;
+		}
+		totalSize = 0;
+
+		for(;;)
+		{
+			numberOfBytes = 0;
+			status = AuxKlibQueryModuleInformation(&numberOfBytes, sizeof(RTL_MODULE_EXTENDED_INFO), 0);
+
+			if (!NT_SUCCESS(status) || !numberOfBytes)
+			{
+				if (!WdfLdrDiags)
+					goto clean;
+				break;
+			}
+			status = RtlULongAdd(numberOfBytes, totalSize, &informationLength);
+
+			if (!NT_SUCCESS(status)) 
+			{
+				if (WdfLdrDiags) 
+				{
+					DbgPrint("WdfLdr: GetImageBase - ");
+					DbgPrint("ERROR: RtlUlongAdd failed with Status 0x%x\n", status);
+				}
+			}
+			else 
+			{
+				numberOfBytes = informationLength;
+			}
+
+			infoBuffer = ExAllocatePoolWithTag(PagedPool, numberOfBytes, WDFLDR_TAG);
+			pModuleInfo = infoBuffer;
+
+			if (pModuleInfo == NULL)
+			{
+				status = STATUS_INSUFFICIENT_RESOURCES;
+
+				if (WdfLdrDiags) 
+				{
+					DbgPrint("WdfLdr: GetImageBase - ");
+					DbgPrint("ERROR: ExAllocatePoolWithTag failed with Status 0x%x\n", status);
+				}
+				goto end;
+			}
+
+			memset(pModuleInfo, 0, numberOfBytes);
+			informationLength = numberOfBytes;
+			status = AuxKlibQueryModuleInformation(&informationLength, sizeof(RTL_MODULE_EXTENDED_INFO), pModuleInfo);
+
+			if (status != STATUS_BUFFER_TOO_SMALL)
+				break;
+
+			ExFreePoolWithTag(pModuleInfo, 0);
+			totalSize += sizeof(RTL_MODULE_EXTENDED_INFO);
+			pModuleInfo = NULL;
```

Comment:
You should nullify infoBuffer here too to avoid a potential double free at line 487.